### ISSUE TITLE
Enable to use of terragrunt instead of terraform

### ISSUE THREAD

**Issue by ctrlaltdev** [infracost/infracost#224]
I always use terragrunt as a wrapper around terraform because I reuse terraform modules between environments or accounts -  and it also allows me to dynamically configure the tf backend and provider I use.

So technically, I don't have a tfdir to provide infracost, as terragrunt only 'compiles' it when I run it.

The good news is most of the terraform commands works with terragrunt (`terragrunt show` for instance)

So I think something like `--use-tg` that would substitute terraform with terragrunt would be great!

**Comment by ctrlaltdev** [comment]
If you want an example of what my IAC looks like with terragrunt - based on terragrunt recommended structure: https://github.com/ctrlaltdev/ctrlalt.dev/tree/main/infra

Under modules, you will find the TF modules, and under config, you will find the terragrunt settings to use those modules.

**Comment by alikhajeh1** [comment]
The following is copied from https://gitlab.com/infracost/infracost-gitlab-ci/-/issues/1#note_448546649 as it's also a recent and relevant request:

> Ayrton: Also, do you know anyone that is using infracost with terragrunt? My new issue is that terragrunt creates separated tfstate for each module (which is good) but for infracost I can only use one tfstate file at time, so the only way I found was creating new pipeline steps for each module which made the pipeline huge (I have more than 20 modules). I'm trying to find a way of using terragrunt for consolidating all the states in a single file just for infracost.

**Comment by ctrlaltdev** [comment]
So maybe mimick terragrunt *-all command behavior

In terragrunt, you can either go in a module and run a command like `terragrunt plan` or you can go to the parent folder that holds multiple modules and run `terragrunt plan-all` and will run terragrunt plan in the modules in parallel.

Maybe infracost could have a similar way of running for terragrunt, that will look recursively in the children folders for terragrunt modules and run against them?

**Comment by robh007** [comment]
With terragrunt you have an option of a pre & post hook, maybe that's an option for infracost to run at either one of those points?

**Comment by ayr-ton** [comment]
@robh007 Hmmm, before apply seems a good one. And it could run for each directory, but probably the output will be messed up. Will try that just for seeing how things will break.

**Comment by robh007** [comment]
Yea @ayr-ton I've had a little play with some code today. 

```
  before_hook "run_infracost" {
    commands = ["apply", "plan"]
    execute = ["infracost"]
  }
```

So the above works but as you say the output is generated per directory, the output is to stdout & doesn't look any different. You could also perhaps output to a json file then use the hcl functions to merge the JSON files together.

The main issue I see is that the TF source is nested within a random directory structure. What you can also do with terragrunt is use a parameter ``download_dir`` & perhaps if infracost could then traverse directories finding .tf files as appose to always expecting the tf files at the top level that may work better. It would also mean that infracost could then use one location to find the TF code as appose to traversing directory structures / parsing hcl files & looking for 'hints' to where all the TF code might be.

**Comment by ayr-ton** [comment]
@robh007 Yeah, the output is uncomprehensive. :-(

**Comment by alikhajeh1** [comment]
@ctrlaltdev and @ayr-ton, just FYI that @robh007, @aliscott, myself and another user were online on [Discord#terragrunt](https://discord.com/invite/rXCTaH3) so we have a spontaneous call to discuss the options. @aliscott is digging into this and we'll post updates here. Feel free to jump into #terragrunt if you prefer to have longer discussions there.

**Comment by aliscott** [comment]
We've added an `infracost report` command which might help with some Terragrunt use cases. It combines multiple infracost JSON files into a single table.

So at the moment for our own infrastructure I can get a table of all the costs by running a script like this:
https://github.com/infracost/infracost/blob/master/scripts/terragrunt/report_all.sh
This runs plan-all, saves all the plan files and then combines them into one table view.

This is now released in v0.7.4. Please try this out and let us know if it can help with your use cases.

**Comment by alikhajeh1** [comment]
Given that we now have https://www.infracost.io/docs/#terragrunt-users, should we close this issue? People can still leave feedback here or open another ticket

**Comment by ayr-ton** [comment]
@alikhajeh1 I liked it a lot. :-) 
Thanks for your time collecting feedback regarding this feature. The infracost report did the thing.
